extra/src/main/java/aima/extra/probability/BigDecimalProbabilityNumber.java

```diff
@@ -0,0 +1,381 @@
+	 * error when applied repeatedly over a sequence of calculations.
+	 */
+	private static final Integer DEFAULT_PRECISION = MathContext.DECIMAL128.getPrecision();
+
+	private static final RoundingMode DEFAULT_PRECISION_ROUNDING_MODE = MathContext.DECIMAL128.getRoundingMode();
+
+	/**
+	 * Maximum integer value that can be accomodated as exponent in the
+	 * BigDecimal pow() function
+	 */
+	private static final int EXPONENT_MAX = 999999999;
+
+	// Internal fields
+
+	private BigDecimal value;
+
+	/**
+	 * MathContext of value.
+	 */
+	private MathContext precisionMathContext = new MathContext(DEFAULT_PRECISION, DEFAULT_PRECISION_ROUNDING_MODE);
+
+	private BigDecimal BIG_DECIMAL_ZERO = new BigDecimal(0, precisionMathContext);
+
+	private BigDecimal BIG_DECIMAL_ONE = new BigDecimal(1, precisionMathContext);
+
+	// Constructors
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from double type.
+	 * BigDecimal.valueOf() builds from a double type based on its canonical
+	 * representation. Precision of double type is at max 15.95 (~15).
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(double value) {
+		if (value < 0 || value > 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(Double.toString(value), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value) {
+		if (null == value || value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(value.toString(), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 * 
+	 * @param mc
+	 *            MathContext to be associated with value.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value, MathContext mc) {
+		precisionMathContext = mc;
+		setConstants();
+		if (value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(value.toString(), precisionMathContext);
+	}
+
+	// Private methods
+
+	/**
+	 * Initialize constants BIG_DECIMAL_ZERO and BIG_DECIMAL_ONE according to
+	 * precisionMathContext that is defined.
+	 */
+	private void setConstants() {
+		BIG_DECIMAL_ZERO = new BigDecimal(0, precisionMathContext);
+		BIG_DECIMAL_ONE = new BigDecimal(1, precisionMathContext);
+	}
+
+	/**
+	 * Covert other implementations of the ProbabilityNumber interface to a
+	 * BigDecimalProbabilityNumber.
+	 * 
+	 * @param that
+	 *            BigDecimalProbabilityNumber.
+	 */
+	private BigDecimalProbabilityNumber toInternalType(ProbabilityNumber that) {
+		if (that instanceof BigDecimalProbabilityNumber) {
```

Comment:
Just a style comment here, your call if you want to adapt. I usually find it helpful to define all public methods first and then private ones at the end of the file. For methods that oreverride/implement an interface order them the same way as they are presented in the interface.